Add validation test for new copyright

willthames opened the pull request:
##### SUMMARY
Ensure new modules without the new copyright header fail
validation

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
validate-modules

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel 52c87c4691) last updated 2017/11/15 13:46:27 (GMT +1000)
  config file = /home/will/src/ansible/ansible.cfg
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible/lib/ansible
  executable location = /home/will/src/ansible/bin/ansible
  python version = 2.7.14 (default, Nov  2 2017, 18:42:05) [GCC 7.2.1 20170915 (Red Hat 7.2.1-2)]
```

Commit:
Update validation test for new copyright

Ensure new modules without the new copyright header fail
validation
Ensure existing modules without copyright in top 20 lines fail

gundalow commented on test/sanity/validate-modules/main.py:
```diff
@@ -384,6 +384,17 @@ def _check_for_gpl3_header(self):
+                    path=self.object_path,
+                    code=108,
+                    msg='GPLv3 simple license header not found: '
+                        'http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#copyright'
```
I wonder if we should direct to the `devel` version of the docs, as this will be for new module development

http://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#copyright

mattclay replied:
Also, all links should be `https` instead of `http`.

mattclay replied:
I'm wondering if we should combine the two GPL license checks into a single more comprehensive test, instead of adding a second one. We could check for:

- Correct license
- Short form (for new modules)
- Correct placement near the top of the file

This would be more thorough than the current substring matching which looks for text strings anywhere in the module.

Also, it appears the logic in the original test is faulty:

```python
        if ('GNU General Public License' not in self.text and
                 'version 3' not in self.text):
```

willthames replied:
@mattclay - it's not apparent to me in what way that logic is faulty.

I've fixed all of your other comments though.

gundalow reviewed (changes requested):
Thanks for doing this, will be very useful

Can you please add `108` into docs/docsite/rst/dev_guide/testing_validate-modules.rst

abadger commented:
It doesn't look like this was addressed:
> Can you please add 108 into docs/docsite/rst/dev_guide/testing_validate-modules.rst

I'm not seeing the flaw in the logic either... @mattclay, should we merge this and then fix up any remaining issues later?

mattclay commented on test/sanity/validate-modules/main.py:
```diff
@@ -375,14 +375,24 @@ def _check_for_sys_exit(self):
+    def _check_gpl3_header(self):
+        header = '\n'.join(self.text.split('\n')[:20])
+        if ('GNU General Public License' not in header and
+                'version 3' not in header):
```
This conditional allows the test to pass if either `GNU General Public License` or `version 3` is present. Shouldn't we require both?

abadger replied:
Ah, I see.  Yeah, should be 
``` python
if not ('GNU General Public License' in header and 'version 3' in header):
```

willthames replied:
Fixed

mattclay commented on test/sanity/validate-modules/main.py:
```diff
@@ -375,14 +375,24 @@ def _check_for_sys_exit(self):
-                msg='GPLv3 license header not found'
+                msg='GPLv3 license header not found in the first 20 lines of the module'
             )
+        if self._is_new_module():
```
I'd restructure the test to be:

```python
if self._is_new_module():
   # new module test here
else:
    # old module test here
```

That way you can't end up with two license errors reported for a new module.

willthames commented:
@abadger the request to document 108 was for the original version of this PR - the current revision doesn't use 108 at all, it reuses 105.

Commit:
Add documentation of 108 error

Create label in developing modules documentation so that
the validation page can point to it

willthames commented:
Error code 108 documented

I've also changed the test so that all modules get a header check, and if that passes, new modules get the new module header check.

mattclay commented on test/sanity/validate-modules/main.py:
```diff
@@ -375,14 +375,24 @@ def _check_for_sys_exit(self):
+    def _check_gpl3_header(self):
+        header = '\n'.join(self.text.split('\n')[:20])
+        if ('GNU General Public License' not in header or
+                'version 3' not in header):
```
This does not match a new style license header such as:

```python
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

mattclay replied:
Yes, the original test was incorrect.

Commit:
Ensure new style copyright header passes test!

willthames commented:
@mattclay - good spot, changed the version matching
